lib/stdlib/src/ms_transform.erl

```diff
@@ -450,10 +448,689 @@ copy_list([],Bound) ->
+  RemainingOperand3 = substitute_promotions_in_guards_expr(Promotions, Operand3),
+  simplify_guard_function(Operator, RemainingOperand1, RemainingOperand2, RemainingOperand3);
+% e.g. is_function/2, '>', etc.
+substitute_promotions_in_guards_expr(Promotions, {tuple, _, [{atom, _, Operator}, Operand1, Operand2]}) ->
+  RemainingOperand1 = substitute_promotions_in_guards_expr(Promotions, Operand1),
+  RemainingOperand2 = substitute_promotions_in_guards_expr(Promotions, Operand2),
+  simplify_guard_function(Operator, RemainingOperand1, RemainingOperand2);
+% e.g. not/1, is_integer/1, etc.
+substitute_promotions_in_guards_expr(Promotions, {tuple, _, [{atom, _, Operator}, Operand]}) ->
+  RemainingOperand = substitute_promotions_in_guards_expr(Promotions, Operand),
+  simplify_guard_function(Operator, RemainingOperand);
+% e.g. '$1'
+substitute_promotions_in_guards_expr(Promotions, {atom,_,AtomInGuard}=Expr) ->
+  case is_column_ref(Expr) of
+    true ->
+      Search =
+        lists:search(fun ({{atom,_,PromotionColumnName},_PromotionValue}) ->
+          PromotionColumnName =:= AtomInGuard
+        end,
+        Promotions),
+      case Search of
+        {value, {_ColumnKey, SubstitutedValue}} -> SubstitutedValue;
+        false -> Expr
+      end;
+    false ->
+      Expr
+  end;
+substitute_promotions_in_guards_expr(_Promotions, Other) ->
+  Other.
+
+% Once we've inlined a column value (e.g. '$1' is replaced with 4, because we saw `$1 =:= 4`),
+% there may be some easy further simplifiations we can make, such as:
+%     {'>', $1, 2}
+%   substituted:
+%     {'>', 4, 2}
+%   simplified:
+%     true
+%
+% Notably, this function doesn't claim to exhaustively
+% partially evaluate guards, and we can't use erl_eval:partial_eval
+% because we have match spec guards, which aren't normal erlang expressions
+simplify_guard_function(is_atom, {atom,_,_}=Val) ->
+  case is_column_ref(Val) of
+    false -> {atom, gen_loc(), true};
+    true -> {tuple, gen_loc(), [{atom, gen_loc(), is_atom}, Val]}
+  end;
+simplify_guard_function(is_float, {float,_,_}) ->
+  {atom,gen_loc(),true};
+simplify_guard_function(is_number, {float,_,_}) ->
+  {atom,gen_loc(),true};
+simplify_guard_function(is_number, {integer,_,_}) ->
+  {atom,gen_loc(),true};
+simplify_guard_function(is_integer, {integer,_,_}) ->
+  {atom,gen_loc(),true};
+simplify_guard_function(is_tuple, {tuple,_,_}) ->
+  {atom,gen_loc(),true};
+simplify_guard_function(is_binary, {bin, _, _}) ->
+  {atom,gen_loc(),true};
+simplify_guard_function('not', {atom, _, true}) ->
+  {atom,gen_loc(),false};
+simplify_guard_function('not', {atom, _, false}) ->
+  {atom,gen_loc(),true};
+simplify_guard_function(Op, Operand) ->
+  {tuple, gen_loc(), [{atom, gen_loc(), Op}, Operand]}.
+
+simplify_guard_function('>', {integer, _, X}, {integer, _, Y}) when X > Y ->
+  {atom,gen_loc(),true};
+simplify_guard_function('>', {integer, _, _}, {integer, _, _}) ->
+  {atom,gen_loc(),false};
```

Comment:
I think this is overly verbose. What about
```
simplify_guard_function('>', {integer, _, X}, {integer, _, Y}) ->
   {atom,gen_loc(), X > Y};
```
or even
```
simplify_guard_function('>', {Num, _, X}, {Num, _, Y}) when Num =:= integer; Num =:= float ->
   {atom,gen_loc(), X > Y};
```

and what about if integers and floats are mixed?
```
simplify_guard_function('>', {integer, _, X}, {float, _, Y})  ->
   {atom,gen_loc(), X > Y};
```

Reply:
I did none of those things, because I wanted follow the pattern from some of the lines above, and in those I was quite conservative, because the pattern not matching doesn't mean the case is definitely false - only that I can't easily prove it is true. I suppose I could tweak some of these cases.

When it came to integers vs. floats, I just didn't want to open that can of worms. I thought I'd try to stay within the realm of things I feel like I largely understand, and potentially refine the optimiser later for more completeness if it appeared to be useful in practice.